pkg/auth/requests/service_account.go

```diff
@@ -0,0 +1,161 @@
+	"net/http"
+	"strings"
+
+	jwtv4 "github.com/golang-jwt/jwt/v4"
+	"github.com/gorilla/mux"
+	authcontext "github.com/rancher/rancher/pkg/auth/context"
+	"github.com/rancher/rancher/pkg/auth/tokens"
+	controllers "github.com/rancher/rancher/pkg/generated/controllers/management.cattle.io/v3"
+	corev1 "github.com/rancher/rancher/pkg/generated/norman/core/v1"
+	mgmtv3 "github.com/rancher/rancher/pkg/generated/norman/management.cattle.io/v3"
+	"github.com/rancher/rancher/pkg/types/config"
+	"github.com/rancher/steve/pkg/auth"
+	"github.com/sirupsen/logrus"
+	v1 "k8s.io/api/authentication/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/apiserver/pkg/endpoints/request"
+	authv1 "k8s.io/client-go/kubernetes/typed/authentication/v1"
+	"k8s.io/client-go/rest"
+)
+
+const (
+	serviceAccountSubjectPrefix = "system:serviceaccount:"
+	settingsObjectName          = "clusterproxyconfig"
+)
+
+type (
+	restConfigGetter  func(cluster *mgmtv3.Cluster, context *config.ScaledContext, secretLister corev1.SecretLister) (*rest.Config, error)
+	authClientCreator func(config *rest.Config) (authv1.AuthenticationV1Interface, error)
+)
+
+// ServiceAccountAuth is an authenticator that authenticates requests using the downstream service account's JWT.
+type ServiceAccountAuth struct {
+	scaledContext             *config.ScaledContext
+	clusterLister             mgmtv3.ClusterLister
+	secretLister              corev1.SecretLister
+	restConfigGetter          restConfigGetter
+	authClientCreator         authClientCreator
+	clusterProxyConfigsGetter controllers.ClusterProxyConfigCache
+}
+
+// NewServiceAccountAuth creates a new instance of ServiceAccountAuth.
+func NewServiceAccountAuth(
+	scaledContext *config.ScaledContext,
+	restConfigGetter restConfigGetter,
+) auth.Authenticator {
+	return &ServiceAccountAuth{
+		scaledContext:    scaledContext,
+		clusterLister:    scaledContext.Management.Clusters("").Controller().Lister(),
+		secretLister:     scaledContext.Core.Secrets("").Controller().Lister(),
+		restConfigGetter: restConfigGetter,
+		authClientCreator: func(config *rest.Config) (authv1.AuthenticationV1Interface, error) {
+			return authv1.NewForConfig(config)
+		},
+		clusterProxyConfigsGetter: scaledContext.Wrangler.Mgmt.ClusterProxyConfig().Cache(),
+	}
+}
+
+// Authenticate the request using the downstream service account's JWT.
+func (t *ServiceAccountAuth) Authenticate(req *http.Request) (user.Info, bool, error) {
+	// Basic checks to see if we even need to proceed
+	info, hasAuth := request.UserFrom(req.Context())
+	if info.GetName() != "system:cattle:error" {
+		return info, hasAuth, nil
+	}
+	clusterID := mux.Vars(req)["clusterID"]
+	if clusterID == "" {
+		return info, hasAuth, fmt.Errorf("no clusterID found in request")
+	}
+	// Check the cluster setting value to determine whether we will continue the auth process
+	settings, err := t.clusterProxyConfigsGetter.Get(clusterID, settingsObjectName)
```

Comment:
I like that approach (List and disable if there is more than one).  If we document this properly, do we really need a webhook check?

Reply:
I like that approach (List and disable if there is more than one).  If we document this properly, do we really need a webhook check?

Reply:
Technically you don't, but the question will be: how do we handle this here? For example, let's say that a ClusterOwner accidentally creates two of these for the same cluster. A project owner then has a service account that makes a request here. We discover there are 2 of these configs. 

At this point we need to return an error (since the information in both configs could conflict and we don't want to get into deciding precedence). However, the SA definitely won't be able to fix the issue, and the user creating/managing the service account may not be able to either, leaving them stuck.

It's for this UX reason that I would recommend the webhook check, but I'm open to other alternatives if you think there's another route to provide a good user experience here.

Reply:
I suppose an ounce of prevention....
Webhook it is